specifyweb/frontend/js_src/lib/components/DataModel/collectionApi.ts

```diff
@@ -85,6 +92,78 @@ export const DependentCollection = Base.extend({
   create: notSupported,
 });
 
+export const IndependentCollection = Base.extend({
```

Comment:
could we extend this class from LazyCollection? or implement some lazy capabilities in it?
that would be pretty important when there are a lot of related records

Reply:
So making the collection lazy is ultimately a backend API problem: we'd need to extend and modify the API to support lazily modifying the collection. 
The complication here is largely how to indicate or determine which related record(s) should be kept in the collection, and which removed. 

The "easiest" implementation I can think of to support this would be to allow dictionaries be passed to independent -to-many relationships with at least two keys: one to indicate related resources which exist in the collection and should be edited or new related resources created, and one to indicate which related resources should be explicitly removed from the collection. 

So something like the following: 

```json
{
    "collectionObjects": {
        "updated": [
            {
                "catalognumber": "#########",
                "text1": "created with the API"
            },
            {
                "id": 123,
                "text1": "updated with the API"
            }
        ],
        "removed": [
            "/api/specify/collectionobject/124"
        ]
    }
}
```

But this approach largely goes against the "intuitive approach" as the currently implemented approach is mostly dependent/independent agnostic: there is no burden on the consumer of the API to know whether the relationship they are working with is independent or not, and they can work with them the same way. 

Of course, we can always support _both_ approaches for independent relationships (and possibly dependent, if we want to keep them unified) or just the "new" approach for independent relationships, but either way this will definitely complicate things. 

Might be worth it though! 😃

Reply:
Realistically we do not have time to develop this new approach for now

Reply:
Would it make sense to approach the API from the opposite direction?
I.e if Agent is used in 2,000 CollectionObjects, and you try to edit those CollectionObjects on the Agent table, realistically user will not change 2,000 CollectionObjects - they will change a few.
So why not simply make separate API requests to update the few CollectionObjects that user changed?
no back-end API changes required

Reply:
mostly transaction support --- everything will happen in a transaction, that's why I proposed an API change (Jason's original implementation did eactly what you mentioned). That way, it's never in an inconsistent state at each save.

As a compromise, what if we, ummm, just not allow removing records, if it currently causes performance issues? that'll simplify loads anyways? If they, really want to bulk remove from to-many side, they can use batch-edit !!

@specify/ux-testing can we please make sure that right now this doesn't cause performance issues? We often talk about "oh, this will damp speed", but never verify that it doesn't (then we have problems like synonym concat on frontend -- which we discussed will happen). Can we test some hot relationships, like agents - co, especially on main agents, and on big databases?

Reply:
> Would it make sense to approach the API from the opposite direction?

Yes, the implementation in #3125 uses that approach, and it should theoretically be much faster as the changes are made far more lazily. 
This API approach _does_ make sure everything happens in a transaction and is consistent: which is defintely a plus. 

> As a compromise, what if we, ummm, just not allow removing records, if it currently causes performance issues? that'll simplify loads anyways? 

Do you mean keep the syntax as is, but treat any records in the array as if they are only new/to be updated? 

Sorry if I caused confusion about where the performance bottleneck here actually is. To be clearer: my comments about record removal from the independent relationship were more about the conceptual/logical complexity rather than performance intensity. 

Record removal _does_ play a role in performance, but record creation/updating is actually more intensive --- likely because the removal of records happen in bulk, while record creation and update is performed on each record as they appear in the array. 
There are actually a few bugs caused _because_ the removal occurs in bulk: 
  - businessrules are not being run
  - version is not being checked/updated 
(I did try to alleviate this somewhat by using bulk=False to [getattr(obj, field_name).remove(*list(to_remove))](https://github.com/specify/specify7/blob/f10e41062107053d082c5a0daa818b54d722ea53/specifyweb/specify/api.py#L690), and businessrules are ran but any changes to the object in `pre_save` signal listeners are not preserved) 
The solution which would ensure the _most_ data integrity for record removal would make all operations comparable to each other. 

Here are some performance metrics (performing each test 3 times and taking the average) I took from the latest commit on this branch: 
https://github.com/specify/specify7/pull/5253/commits/f10e41062107053d082c5a0daa818b54d722ea53

It is measured in time, not ideal, but should give a rough idea on how the operations on records associated with the relationship scale. 

| Test                    | Real Time (Seconds)                                                    | CPU Time (Seconds)                                                     |
|-------------------------|------------------------------------------------------------------------|------------------------------------------------------------------------|
| Adding 200 Existing COs[^1]          | 63.895977445994504                                                     | 27.936980204                                                           |
| Removing 200 COs[^2]        | 4.510533169028349                                                      | 2.0929900000000004                                                     |
| Changing 1 CO of 200[^3]    | 6.256301502988208                                                      | 3.1918773569999956                                                     |
| Removing 100 of 200 COs | Add/Update Time:  1.7626766580000037  Remove Time:  0.5865840419999984 | Add/Update Time:  1.7626766580000037  Remove Time:  0.5865840419999984 |
| Adding 500 Existing COs          | 134.89527818595525                                                     | 56.609880032                                                           |
| Changing 1 CO of 500    | 14.87152375705773                                                      | 7.410588650000008                                                      |
| Removing 250 of 500 COs | Add/Update Time:  7.591786962002516  Remove Time:  6.227154252002947   | Add/Update Time:  3.859158597000004  Remove Time:  2.6677741320000052  |
| Removing 500 COs        | 12.055516088963486                                                     | 5.102607159999991                                                      |

[^1]: This type of test is to measure how long the update operation on each CollectionObject takes
[^2]: Some type of test as [^1], but the operation is performed in bulk
[^3]: All other COs are passed as an API link. Specify performs a fetch on each record and checks whether they are associated with the related or not, and if so, skips the update operation

Reply:
Think more simply. If we don’t allow deletion, we don’t need to fetch all, right? So, there could be a million co, and we fetch just 20, similar to record selector? So yes, you got what I meant. 

also, version update can be fixed by just bumping version. You could even just increment it using an F() bc you already got a lock on them (so you don’t need to get all of them and separately increment version — performance optimization).

Reply:
Also, if you open andy with a catalogerof, it doesn’t even load.

Reply:
Also, the problem of business rules can be simply solved by explicitly updating the remote object?? Why are you sticking to remove if it causes problems?

Reply:
> Realistically we do not have time to develop this new approach for now

@CarolineDenis 
The way the feature is implemented right now, it can crash the server or the freeze the client because payload size is going to be just wayyy to large in some circumstances.

---

@melton-jason 

> But this approach largely goes against the "intuitive approach" as the currently implemented approach is mostly dependent/independent agnostic: there is no burden on the consumer of the API to know whether the relationship they are working with is independent or not, and they can work with them the same way.
>
> Of course, we can always support both approaches for independent relationships (and possibly dependent, if we want to keep them unified) or just the "new" approach for independent relationships, but either way this will definitely complicate things.

I like this implementation.

To be fair, knowing whether relationship is dependent or not is already required since (until now) you would get an exception if you try to include independent objects in the API request.

Another possibility:

- The GET requests still don't provide independent resources - front-end fetches these lazily
- Have a PUT/POST API endpoint that accepts an array of resources - it will be equivalent to calling the endpoint several times except the payload within a single request is part of the same transaction.

---

> There are actually a few bugs caused because the removal occurs in bulk:
> businessrules are not being run
> version is not being checked/updated

Yeah, this is part of the reason why I went with the approach of requiring a separate "Save" button for independent resources - no issues with business rules not being run, or save blockers being misapplied to a wrong resource.  And no need for API changes. And since user has to click the "Save" button for each resource (yes, more annoying for the user), specify is not on the hook as much for ensuring everything happens as part of a single transaction.

---

@realVinayak 

> How in the world is objects more than 200+ ever be touched by the user -- adding or saving? Stop creating unnecessary problems and writing complicated code to solve it (you haven't solved any), when there is a better place for it (batch-edit).

Agree. So maybe the solution is to make independent subviews readonly, but add an easy way to get from here to the batch-editor (maybe via a new button in the form-meta menu, or via a new button next to where the +/-/search buttons are)

Reply:
> specify is not on the hook as much for ensuring everything happens as part of a single transaction.

Nice idea Max! Implementation wise, we could even prevent editing independent if main resource needs saved (or has a save blocker), and then for each edit to independent, we refresh the entire page? Sounds like a simple way, given that even in failure, only change to 1 independent will be lost. I'd much rather prefer a way that just gets it done in the most minimal way.

> So maybe the solution is to make independent subviews readonly, but add an easy way to get from here to the batch-editor 

Need to look into this, but as long as a query builder query can be constructed, batch-edit can be done (at least in the way we want to, in this case, automatcially just from the form). Otherwise, they can always construct a query manually.